File: backend/src/schemas/billing.py
"""Pydantic schemas for the Licensing / Billing domain.

Two read surfaces:
- ``BillingSummary`` — a single tenant's current plan + live usage + invoice
  stub (``GET /v1/billing/summary``).
- ``PlanRead`` / ``PlanList`` — the global plan catalog for the operator Console
  (``GET /v1/admin/plans``).

Money note: plans store an integer ``price_cents`` server-side; these schemas
expose ``price`` already converted to the major unit (reais) as a float so the
frontend's ``fmtBRL`` can format it directly.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel

from models.enums import SubscriptionStatus
from models.plan import Plan


class PlanRead(BaseModel):
    """A plan in the global catalog. ``None`` limits mean *unlimited*."""

    id: uuid.UUID
    slug: str
    name: str
    tagline: str | None
    price: float  # major unit (reais), converted from price_cents
    currency: str
    max_members: int | None
    max_orders_per_month: int | None
    max_integrations: int | None
    max_storage_gb: int | None
    is_public: bool
    sort_order: int
    active: bool

    @classmethod
    def from_plan(cls, plan: Plan) -> PlanRead:
        return cls(
            id=plan.id,
            slug=plan.slug,
            name=plan.name,
            tagline=plan.tagline,
            price=plan.price_cents / 100,
            currency=plan.currency,
            max_members=plan.max_members,
            max_orders_per_month=plan.max_orders_per_month,
            max_integrations=plan.max_integrations,
            max_storage_gb=plan.max_storage_gb,
            is_public=plan.is_public,
            sort_order=plan.sort_order,
            active=plan.active,
        )


class PlanList(BaseModel):
    items: list[PlanRead]
    total: int


class UsageMetric(BaseModel):
    """One usage dimension measured against the active plan's limit.

    ``limit is None`` → unlimited. ``tracked=False`` → Orion does not yet model
    this dimension (integrations, storage), so ``used`` is a placeholder and the
    UI should render it as "not tracked" rather than a real number.
    """

    key: str  # "members" | "orders_month" | "integrations" | "storage"
    used: int
    limit: int | None
    tracked: bool


class InvoiceStub(BaseModel):
    """Placeholder invoice shape. Payments are out of scope — the list is always
    empty for now, but the field exists so the contract is stable."""

    id: str
    period: str
    amount: float
    currency: str
    status: str


class SubscriptionRead(BaseModel):
    """The company's resolved subscription. ``persisted=False`` when the company
    has no Subscription row and is defaulted onto the free plan."""

    status: SubscriptionStatus
    period_start: datetime | None
    period_end: datetime | None
    cancel_at: datetime | None
    persisted: bool


class BillingSummary(BaseModel):
    plan: PlanRead
    subscription: SubscriptionRead
    usage: list[UsageMetric]
    invoices: list[InvoiceStub]

File: frontend/src/lib/schemas/billing.ts
/**
 * TypeScript types for the Licensing / Billing domain.
 * Mirrors the Pydantic shapes in `backend/src/schemas/billing.py`.
 *
 * Money: the backend already converts cents → reais, so `price` is a number in
 * the major unit (BRL) ready for `fmtBRL`. Monthly limits are `number | null`
 * where `null` means unlimited.
 */

export const SUBSCRIPTION_STATUSES = [
  "active",
  "trialing",
  "past_due",
  "paused",
  "cancelled",
  "free",
] as const;

export type SubscriptionStatus = (typeof SUBSCRIPTION_STATUSES)[number];

/** A usage dimension measured against the active plan's limit. */
export type UsageMetricKey = "members" | "orders_month" | "integrations" | "storage";

export type UsageMetric = {
  key: UsageMetricKey;
  used: number;
  /** `null` means unlimited. */
  limit: number | null;
  /** `false` means Orion doesn't model this dimension yet (render "not tracked"). */
  tracked: boolean;
};

export type PlanRead = {
  id: string;
  slug: string;
  name: string;
  tagline: string | null;
  /** Major unit (reais), already converted from price_cents. */
  price: number;
  currency: string;
  max_members: number | null;
  max_orders_per_month: number | null;
  max_integrations: number | null;
  max_storage_gb: number | null;
  is_public: boolean;
  sort_order: number;
  active: boolean;
};

export type PlanList = { items: PlanRead[]; total: number };

export type SubscriptionRead = {
  status: SubscriptionStatus;
  period_start: string | null;
  period_end: string | null;
  cancel_at: string | null;
  persisted: boolean;
};

export type InvoiceStub = {
  id: string;
  period: string;
  amount: number;
  currency: string;
  status: string;
};

export type BillingSummary = {
  plan: PlanRead;
  subscription: SubscriptionRead;
  usage: UsageMetric[];
  invoices: InvoiceStub[];
};
